packages/gatsby/src/db/loki/nodes-query.js

```diff
@@ -231,6 +242,24 @@ function toSortFields(sortArgs) {
   ])
 }
 
+// Every time we run a query, we increment a counter for each of its
+// fields, so that we can determine which fields are used the
+// most. Any time a field is seen more than `FIELD_INDEX_THRESHOLD`
+// times, we create a loki index so that future queries with that
+// field will execute faster.
```

Comment:
I think this is fine. Something to think about in the future is that we can analyze queries (through query AST) when we are extracting / compiling them and then we can know what fields are used for filtering. We will know how many times we will run queries (number of pages using same template/query). Not sure if it's worth extra engineering to do that tho.

Reply:
Yeah, I thought about this approach too but didn't want to get lost in the weeds of query ASTs. I definitely think it's the right approach in the future. Thankfully the CPU overhead of the above is very minimal.